ext/memcached_rb/memcached_rb.c

```diff
@@ -1,460 +1,518 @@
+	"UnknownStatKey", // MEMCACHED_UNKNOWN_STAT_KEY
+	"E2big", // MEMCACHED_E2BIG
+	"InvalidArgument", // MEMCACHED_INVALID_ARGUMENTS
+	"KeyTooBig", // MEMCACHED_KEY_TOO_BIG
+	"AuthProblem", // MEMCACHED_AUTH_PROBLEM
+	"AuthFailure", // MEMCACHED_AUTH_FAILURE
+	"AuthContinue", // MEMCACHED_AUTH_CONTINUE
+	"ParseError", // MEMCACHED_PARSE_ERROR
+	"ParseUserError", // MEMCACHED_PARSE_USER_ERROR
+	"Deprecated", // MEMCACHED_DEPRECATED
+	"InProgress", // MEMCACHED_IN_PROGRESS
+	"ServerTemporarilyDisabled", // MEMCACHED_SERVER_TEMPORARILY_DISABLED
+	"ServerMemoryAllocationFailure", // MEMCACHED_SERVER_MEMORY_ALLOCATION_FAILURE
+	"MaximumReturn", // MEMCACHED_MAXIMUM_RETURN
+	"ConnectionSocketCreateFailure", // MEMCACHED_CONNECTION_SOCKET_CREATE_FAILURE
 };
-#define MEMCACHED_ERROR_COUNT 51
 
-VALUE rb_eMemcachedErrors[MEMCACHED_ERROR_COUNT];
+#define MEMCACHED_ERROR_COUNT (ARRAY_SIZE(MEMCACHED_ERROR_NAMES))
+static VALUE rb_eMemcachedErrors[MEMCACHED_ERROR_COUNT];
 
 void
-handle_memcached_return(memcached_return_t rc)
+rb_memcached_error_check(memcached_return_t rc)
 {
-  if (rc == MEMCACHED_SUCCESS)
-    return;
-
-  VALUE rb_error, boom;
-  const char * message;
-  if (rc > 0 && rc < MEMCACHED_ERROR_COUNT) {
-    rb_error = rb_eMemcachedErrors[rc];
-    if(Qnil == rb_error)
-      return;
-    message = memcached_strerror(NULL, rc);
-  } else {
-    rb_error = rb_eMemcachedError;
-    message = "Memcached returned type not handled";
-  }
-
-  boom = rb_exc_new2(rb_error, message);
-  rb_exc_raise(boom);
-}
-
-#define Wrap_return_t(rc) \
-{ \
-  handle_memcached_return(rc); \
-  if(rc == MEMCACHED_BUFFERED) \
-    return Qnil; \
-  else \
-    return rc == MEMCACHED_SUCCESS ? Qtrue : Qfalse; \
+	VALUE rb_err;
+	const char * message;
+
+	if (rc == MEMCACHED_SUCCESS)
+		return;
+
+	if (rc > 0 && rc < MEMCACHED_ERROR_COUNT) {
+		rb_err = rb_eMemcachedErrors[rc];
+		if (NIL_P(rb_err))
+			return;
+		message = memcached_strerror(NULL, rc);
+	} else {
+		rb_err = rb_eMemcachedError;
+		message = "Memcached returned type not handled";
+	}
+
+	rb_exc_raise(rb_exc_new2(rb_err, message));
 }
 
-memcached_ctx *
-get_ctx(VALUE obj)
-{
-  memcached_ctx* ctx;
-  Data_Get_Struct(obj, memcached_ctx, ctx);
-  return ctx;
-}
+#define rb_memcached_return(rc) { \
+		rb_memcached_error_check(rc); \
+		switch (rc) { \
+		case MEMCACHED_BUFFERED: return Qnil; \
+		case MEMCACHED_SUCCESS: return Qtrue; \
+		default: return Qfalse; }}
 
-static void
-free_memc(memcached_ctx * ctx)
-{
-  memcached_free(ctx->memc);
-}
 
 static VALUE
-allocate(VALUE klass)
+rb_memcached_alloc(VALUE klass)
 {
-  memcached_ctx * ctx;
-  VALUE rb_obj = Data_Make_Struct(klass, memcached_ctx, NULL, free_memc, ctx);
-
-  ctx->memc = memcached_create(NULL);
-  return rb_obj;
+	memcached_st *mc = memcached_create(NULL);
+	return Data_Wrap_Struct(klass, NULL, memcached_free, mc);
 }
 
 static VALUE
-rb_connection_initialize(VALUE self, VALUE rb_servers)
+rb_connection_new(VALUE klass, VALUE rb_servers)
```

Comment:
I've removed the initialize method altogether. A common idiom for wrappers that create a structure in C-land is to skip the allocation function + initialization function and instead provide a `YourClass#new` as a singleton, which returns directly the new object. Saves a lot of complexity!